syntax = "proto3";

option go_package = "github.com/numaproj/numaflow-go/pkg/apis/proto/flatmap/v1";

import "google/protobuf/empty.proto";
import "google/protobuf/timestamp.proto";

package flatmap.v1;

service Flatmap {
  // MapFn applies a function to each map request element.
  rpc MapFn(stream MapRequest) returns (stream MapResponse);

  // IsReady is the heartbeat endpoint for gRPC.
  rpc IsReady(google.protobuf.Empty) returns (ReadyResponse);
}

/**
 * MapRequest represents a request element.
 */
message MapRequest {
  repeated string keys = 1;
  bytes value = 2;
  google.protobuf.Timestamp event_time = 3;
  google.protobuf.Timestamp watermark = 4;
  map<string, string> headers = 5;
  string uuid = 6;
}

/**
 * MapResponse represents a response element.
 */
message MapResponse {
  message Result {
    repeated string keys = 1;
    bytes value = 2;
    repeated string tags = 3;
    string uuid = 4;
    bool EOR = 5;
  }
  Result result = 1;
}

/**
 * ReadyResponse is the health check result.
 */
message ReadyResponse {
  bool ready = 1;
}